igvc/src/cnn/cnn.py

```diff
@@ -0,0 +1,173 @@
+        cam_transform_translation, cam_transform_rotation = transform_listener.lookupTransform('/base_footprint', '/optical_cam_center', rospy.Time(0))
+        self.cam_transform_rotation_matrix = ros_tf.transformations.quaternion_matrix(cam_transform_rotation)[:-1,:-1]
+        self.cam_transform_translation = np.asarray(cam_transform_translation)
+        print self.cam_transform_translation
+        print self.cam_transform_rotation_matrix
+
+        self.init_point_cloud_array()
+
+        self.im_publisher = rospy.Publisher(publisher_topic, Image, queue_size=1)
+        self.cloud_publisher = rospy.Publisher("/semantic_segmentation_cloud", PointCloud2, queue_size=1)
+        self.subscriber = rospy.Subscriber(subscriber_topic, Image, self.image_callback, queue_size=1, buff_size=10**8)
+
+
+
+
+
+    def load_graph(self, frozen_graph_filename):
+        # We load the protobuf file from the disk and parse it to retrieve the
+        # unserialized graph_def
+        with tf.gfile.GFile(frozen_graph_filename, "rb") as f:
+            graph_def = tf.GraphDef()
+            graph_def.ParseFromString(f.read())
+
+        # Then, we import the graph_def into a new Graph and returns it
+        with tf.Graph().as_default() as graph:
+            # The name var will prefix every op/nodes in your graph
+            # Since we load everything in a new graph, this is not needed
+            tf.import_graph_def(graph_def, name='')
+        return graph
+
+
+
+    def init_point_cloud_array(self):
+        world_points_array = np.empty( (self.resize_height,self.resize_width) ,dtype=object)
+        for r,c in np.ndindex(self.resize_height,self.resize_width):
+            ray = np.asarray( self.camera_model.projectPixelTo3dRay( (c,r) ) )
+            ray_in_world = np.matmul( self.cam_transform_rotation_matrix, ray )
+            scale = -self.cam_transform_translation[2] / ray_in_world[2]
+            world_point = scale * ray_in_world + self.cam_transform_translation
+            world_points_array[r,c] = (world_point[0], world_point[1], world_point[2])
+        self.world_point_array = world_points_array
+
+
+    def image_callback(self, image_msg):
+        start = timer()
+        try:
+            cv_image = self.bridge.imgmsg_to_cv2(image_msg, "bgr8")
+        except CvBridgeError as e:
+            print(e)
+        cv_image = cv2.cvtColor(cv_image, cv2.COLOR_BGR2RGB)
+        cv_image = cv2.resize(cv_image, (self.resize_width, self.resize_height))
+        im = np.asarray(cv_image)
+
+        image_pl = self.graph.get_tensor_by_name('Placeholder_1:0')
+        softmax = self.graph.get_tensor_by_name('Validation/decoder/Softmax:0')
+
+        output = self.sess.run([softmax], feed_dict={image_pl: im})
+        shape = im.shape
+        output_image = output[0][:, 1].reshape(shape[0], shape[1])
+
+        threshold = 0.5
+        im_threshold = output_image > threshold
+        world_points = self.world_point_array[im_threshold].tolist()
```

Comment:
We convert this numpy array to a regular list, then convert it back on line 151.